### PR TITLE
Don't re-alias top-level type aliases with local type aliases

### LINKED ISSUE #43622: React HOCs generate extremely large and incorrect .d.ts declaration files, regression from 4.1.3
# Bug Report

While trying to upgrade from 4.1.3 to 4.2.3 in our monorepo, we ran into a regression that's preventing us from upgrading at the moment. (This is the same monorepo mentioned in https://github.com/microsoft/TypeScript/issues/41570 cc @brieb @lencioni) 

When using React HOCs and letting TS infer returned types, the type inferred and included in the corresponding `.d.ts` is:
- much larger, with a large number of conditional types
- incorrect, in some instances, removing wrapped component props optionality
- inconsistent with direct type checking (leading to the language server reporting no errors but `tsc` reporting errors, in some cases)
- much slower during compilation 

### 🔎 Search Terms

HOC

Declaration mismatch

Upgrade

Performance regression

### 🕗 Version & Regression Information

4.2.* seems to exhibit the issue. A git bisect with locally built compilers seems to point to 34f0e32dc667eb8c7b52b3fc4067316b4ae97774 as the first bad commit (however, despite the title of the commit, the npm package for `4.2.1` also has the issue).

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAJQKYEMDG8BmUIjgcilQ3wG4AocgegCoby4a4BxJGAZzhgAsk4wcYTpmhwUiYvCQAbJCCQA7eDACeYJABo4Ad2A8IAV3gCk7JFABuwBQHMuvOBDAxgEBSml6VjzHAAmSJgoBtIwAAqC7AB0DPSMcACSCuxgwER+cABGKgBccNwwMEI5VFQ2etwGmVFouFRBaEiZEBAA1vXSENpUmZ2ZPQCMAKwADChIAGwAHADMKCgzUxPjAOwATACcE35DftuZGyszACxDx0gDQStoKGtU2k3sekhUilR+EGjsVEToMFSqdTREB+ADEvwwMjkiiKkUBSAgmAAtLVwG4YUjYrE4AA5CAwJB5ABiHmknEy6FaXAgYgU3hEUDgqLAKBcmWAnlUOgqcAAUgBlAAaTNwkAUMOisSo5HhcAAorJ5EoIk52AAeADCAD44ABeOAauBIAAeBIUfk4yD+UQ1ovRSgAKmokGqUHStQw4AB+A1G02KC1wADecHF2jgAAoBKq8tZMOY4GEtLUlCaYF68m6VABKTN0uAAX09cG9ieLcDyhrTAc4UcisYU8cZYTgADJg0zuBy-EQFBmJNarRg8QFC8m3ATTf2s9m9Tqs+WfWFy3mVJ7VxRqHRsQAZYCtPgK6HKyJq+GIg128VKLVaHjAThGDleOA2NicHh8WUX8TM+0wfBOGjIQNGxTIjH4IgzEsUx7D4JwXDcDwXwvAIghCcJIhiRhsSSFI0iQDJsjyAoinYEoygqKoajqBomhadpME6bpeggfpMmGMZJlmeZFmWJB1i2HY9j8A4jlOc5LhQa5bnuR5nleBR3k+b4IX+eFgTBNSoSVGAU0wYAbHPZE-2vGBMUYaVZSPXTbUbQzNR1fUBUFKI90yKAUCgFQAFk3RQN8-AAQUKKBgHAgl1Q1LQbJhFUhEcrVNxNSBYDgawCSgOi4AAdQqflcDYLtbHizgg09dhCvvWw8kAqripsMhyCLcgUugLADAUDBXAUbkeAK+RqpsTU-TNQMhxgG0rxhJ11Fdd0tQjJJxSgW00TMytZ3KksS1lNaxTiyI9XlRUYTsgzhu1Cgdq4Z1EyO-UAHkQD0TVppPVUtAPFQLzy-r6usGxSqSygbpTdh4ByzywHUPx9v-Y60GkFB2E4P7uAGorAfhszRprAcMCm9aYTVYHg3LEtewCKAIy2imdqIGADCgXq1WW8wcZhYMoh5+9omAzgUcvYmPqEAs4EqwaGt1AAiSWsdsGW4CoEGbpLIsbpam7SgNN0sj4NBfgJDIhfEFyjVOpRy3BiBZCiTobAjNUoZQGHCM5pRuZ5oNxdNulxZV7NNxuxnmd6l23bh96YAoIsgA)

Note that your browser may slow down significantly. Check the `.d.ts` tab. Compare with version 4.1.3.

This does not fully reproduce the error because of cross project dependencies (the type is not only large but also incorrect).

### 💻 Code

To reproduce the issue in more details, check https://github.com/ms/ts-hoc-repro (to reproduce: `npm i && npx tsc -p projectA/ && npx tsc -p projectB/`). The two issues:
- `projectA/withSomething.d.ts` is very large, as seen on the playground, see https://github.com/ms/ts-hoc-repro/blob/master/projectA/withSomething.d.ts
- `projectB/` won't type check because it complains about a missing prop, but that prop is optional in the inner component (so the HOC return type is incorrect)

### 🙁 Actual behavior

- Incorrect type for wrapped components
- Large `.d.ts`
- Slow compilation using `composite: true` because of the large `.d.ts`

### 🙂 Expected behavior

- Correct type for wrapped component
- `.d.ts` files equivalent to 4.1.3
- No speed regression on compilation time

Last, I ran `tsc` with `--generateTrace` for 4.1 and 4.2. Here's a screenshot for 4.2 with `transformNodes`, in the emit phase, taking multiple seconds (this call is invisible for 4.1)

image

### Workaround

We've been able to mitigate this issue by explicitly annotating the return type of the HOC (using the exact same generic helpers, `ElementConfig`, etc.)

### PR DESCRIPTION
**Pull request by ahejlsberg**
Fixes #43622.

### CODE REVIEW

**Comment by ahejlsberg** [comment]
I did not add the repro from #43622 as a test since it requires React and other libraries. I have manually verified that #43622 is fixed by this PR, and we already have a test that verifies the new local type alias behavior (see the baseline changes).

File: tests/baselines/reference/conditionalTypes1.errors.txt [hunk 1]
```diff
@@ -50,7 +50,7 @@ tests/cases/conformance/types/conditional/conditionalTypes1.ts(159,5): error TS2
 tests/cases/conformance/types/conditional/conditionalTypes1.ts(160,5): error TS2322: Type 'T' is not assignable to type 'ZeroOf<T>'.
   Type 'string | number' is not assignable to type 'ZeroOf<T>'.
     Type 'string' is not assignable to type 'ZeroOf<T>'.
-tests/cases/conformance/types/conditional/conditionalTypes1.ts(263,9): error TS2403: Subsequent variable declarations must have the same type.  Variable 'z' must be of type 'T1', but here has type 'T2'.
+tests/cases/conformance/types/conditional/conditionalTypes1.ts(263,9): error TS2403: Subsequent variable declarations must have the same type.  Variable 'z' must be of type 'T1', but here has type 'Foo<T & U>'.
```
Comment (amcasey) [inline tests/baselines/reference/conditionalTypes1.errors.txt:53]: This seems worse, but I believe I understood you to say it's just a return to our old behavior?

**Review by amcasey — APPROVED** [review]
This code appears to match the explanation you gave at the design meeting, but I don't understand the feature well enough to predict possible consequences.  Is there anything specific you wanted me to test/check?

File: tests/baselines/reference/conditionalTypes1.errors.txt (hunk 1 above)
Comment (ahejlsberg) [reply tests/baselines/reference/conditionalTypes1.errors.txt:53]: Yes, it's a return to the old behavior, the rationale being that `Foo<xxx>` is accessible outside the function whereas `T2` is not. I think very little code will be affected by this change, as it is relatively rare to have local type aliases.

**Comment by ahejlsberg** [comment]
> Is there anything specific you wanted me to test/check?

Not really. I suppose the debatable part is whether we just want to point users at the workaround (i.e. changing `xxx` to `[xxx][0]`) and not actually merge this PR.

**Comment by amcasey** [comment]
> > Is there anything specific you wanted me to test/check?
> 
> Not really. I suppose the debatable part is whether we just want to point users at the workaround (i.e. changing `xxx` to `[xxx][0]`) and not actually merge this PR.

Is there some way we could supplement tracing to make it clear when a bad expansion happens so that we can at least point users to the relevant type declarations?  If we can't assist users in finding _where_ to apply that workaround, I'd be pretty reluctant to recommend they do so.

**Comment by DanielRosenwasser** [comment]
@typescript-bot perf test this faster

**Review by DanielRosenwasser — APPROVED** [review]

**Comment by ahejlsberg** [comment]
> Is there some way we could supplement tracing to make it clear when a bad expansion happens so that we can at least point users to the relevant type declarations?

I can't think of any. If we preserved all possible type aliases then perhaps, but in that case we'd probably just add logic to pick the first accessible alias. But as we discussed in the design meeting, I don't think the additional complexity is merited.

Overall, I think we're better off with the behavior in this PR so I'll merge it.

**Pull request merged**